File: python/array/leetcode/container_with_most_water.py
from typing import List

# 装水最多的容器


class Solution:
    # 暴力法，两次遍历寻找最大值
    def maxArea1(self, height: List[int]) -> int:
        res = 0
        for i in range(len(height) - 1):
            for j in range(i + 1, len(height)):
                area = min(height[i], height[j]) * (j - i)
                if area > res:
                    res = area
        return res

    # 双指针法
    def maxArea2(self, height: List[int]) -> int:
        res, i, j = 0, 0, len(height) - 1
        while i < j:
            res = max(res, min(height[i], height[j]) * (j - i))
            if height[i] < height[j]:
                i += 1
            else:
                j -= 1
        return res

File: go/array/leetcode/container_with_most_water.go
package main

//装水最多的容器

//暴力法，两层循环
func maxArea1(height []int) int {
	max := 0
	for i := 0; i < len(height)-1; i++ {
		for j := i + 1; j < len(height); j++ {
			res := (j - i) * min(height[i], height[j])
			if res > max {
				max = res
			}
		}
	}

	return max
}

func min(x int, y int) int {
	if x > y {
		return y
	}

	return x
}

//双指针法
func maxArea2(height []int) int {
	max, i, j := 0, 0, len(height)-1
	for i < j {
		res := (j - i) * min(height[i], height[j])
		if res > max {
			max = res
		}

		if height[i] < height[j] {
			i++
		} else {
			j--
		}
	}

	return max
}
